src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -591,3 +646,288 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+
+	if (parentState->parentIsBranch)
+	{
+		if (containsVar)
+		{
+			parentState->subtreesWithVar = lappend(parentState->subtreesWithVar, copy);
+			parentState->subtreeOrder = lappend_int(parentState->subtreeOrder, 0);
+		}
+		else
+		{
+			parentState->subtreesWithoutVar = lappend(parentState->subtreesWithoutVar,
+													  copy);
+			parentState->subtreeOrder = lappend_int(parentState->subtreeOrder, 1);
+		}
+	}
+
+	/* never set containsVar=false because parentState is shared between sibling trees */
+	if (containsVar)
+	{
+		parentState->containsVar = true;
+	}
+
+	if (isBranch && parentState->isTopLevel &&
+		list_length(childState.subtreesWithVar) == 0)
+	{
+		return EvaluateExpression(expression);
+	}
+
+	if ((list_length(childState.subtreesWithVar) != 0) &&
+		(list_length(childState.subtreesWithoutVar) != 0))
+	{
+		if (IsA(expression, FuncExpr))
+		{
+			((FuncExpr *) copy)->args = RewriteArgs(&childState);
+		}
+
+		if (IsA(expression, OpExpr))
+		{
+			((OpExpr *) copy)->args = RewriteArgs(&childState);
+		}
+	}
+
+	return copy;
+}
+
+/* evaluate every non-Var arg and return an argument list */
+static List * RewriteArgs(ExpressionWalkerState *state)
+{
+	List *rewrittenArgs = NULL;
+	ListCell *nextArgCell = NULL;
+	int nextArg = 0;
+
+	foreach(nextArgCell, state->subtreeOrder)
+	{
+		nextArg = lfirst_int(nextArgCell);
+		if (nextArg)
+		{
+			Node *original = (Node *)linitial(state->subtreesWithoutVar);
+			Node *evaluated = EvaluateExpression(original);
+			rewrittenArgs = lappend(rewrittenArgs, evaluated);
+			state->subtreesWithoutVar =
+				list_delete_first(state->subtreesWithoutVar);
+		}
+		else
+		{
+			rewrittenArgs = lappend(rewrittenArgs,
+									linitial(state->subtreesWithVar));
+			state->subtreesWithVar = list_delete_first(state->subtreesWithVar);
+		}
+	}
+
+	return rewrittenArgs;
+}
+
+static Node * EvaluateExpression(Node *expression)
+{
+	if(IsA(expression, FuncExpr))
+	{
+		FuncExpr *expr = (FuncExpr *)expression;
+
+		return (Node *) citus_evaluate_expr((Expr *) expr,
+									  expr->funcresulttype,
+									  exprTypmod((Node *) expr),
+									  expr->funccollid);
+	}
+
+	if(IsA(expression, OpExpr))
+	{
+		OpExpr *expr = (OpExpr *)expression;
+
+		return (Node *) citus_evaluate_expr((Expr *) expr,
+									  expr->opresulttype,
+									  exprTypmod((Node *) expr),
+									  expr->opcollid);
+	}
+
+	// TODO: What else should we evaluate here?
+	// What else can contain a function call?
+	return expression;
+}
+
+
+static void InitializeWalkerState(ExpressionWalkerState *state)
+{
+	state->isTopLevel = false;
+	state->containsVar = false;
+	state->parentIsBranch = false;
+	state->subtreesWithVar = NULL;
+	state->subtreesWithoutVar = NULL;
+	state->subtreeOrder = NULL;
+}
+
+/*
+ * pre-evaluate a constant expression, a copy of pg's evaluate_expr
+ *
+ * We use the executor's routine ExecEvalExpr() to avoid duplication of
+ * code and ensure we get the same result as the executor would get.
+ */
```

Comment:
Should note that this is a more or less direct copy of evaluate_expr(). Probably needs /* INDENT OFF */ markers to survive citus_indent.